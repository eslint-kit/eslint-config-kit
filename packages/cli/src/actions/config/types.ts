import { Config } from '../../lib/shared-types'

export interface Answers {
  updatedConfigs: Config[]
  addedConfigs: Config[]
  deletedConfigs: Config[]
  shouldAddRecommendedPrettierConfig?: boolean
}
