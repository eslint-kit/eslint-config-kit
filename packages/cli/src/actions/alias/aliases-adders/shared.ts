/* eslint-disable @typescript-eslint/no-explicit-any */

import merge from 'deepmerge'

export const combineMerge = (target: any, source: any, options: any): any[] => {
  const destination = target.slice()

  source.forEach((item: any, index: number) => {
    if (typeof destination[index] === 'undefined') {
      destination[index] = options.cloneUnlessOtherwiseSpecified(item, options)
    } else if (options.isMergeableObject(item)) {
      destination[index] = merge(target[index], item, options)
    } else if (target.indexOf(item) === -1) {
      destination.push(item)
    }
  })

  return destination
}
