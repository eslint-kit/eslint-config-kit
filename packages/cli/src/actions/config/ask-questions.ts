import inquirer from 'inquirer'
import chalk from 'chalk'
import { Config, PrettierConfigMeta, Choice } from '../../lib/shared-types'
import { toMap } from '../../lib/util/to-map'
import { Answers } from './types'

interface ConfigChoice extends Choice {
  value: Config
}

const DEFAULT_CONFIG_CHOICES: ConfigChoice[] = [
  { name: 'Base (required)', value: 'base', checked: true },
  { name: 'Prettier', value: 'prettier' },
  { name: 'React', value: 'react' },
  { name: 'Node', value: 'node' },
  { name: 'TypeScript', value: 'typescript' },
]

const configToNameMapper: Record<Config, string> = {
  'base': 'Base',
  'prettier': 'Prettier',
  'react': 'React',
  'react-new-jsx-transform': 'React With New JSX Transform Support',
  'node': 'Node',
  'typescript': 'TypeScript',
}

function configToName(configKey: Config): string {
  return configToNameMapper[configKey]
}

function createConfigsMessage({
  hasInstalledConfigs,
}: {
  hasInstalledConfigs: boolean
}): string {
  let result = 'Choose configs'

  if (hasInstalledConfigs) {
    result += ' '
    result += '(uncheck installed configs to remove them)'
  }

  result += ':\n'

  return result
}

function createConfirmationMessage({
  addedConfigs,
  deletedConfigs,
  remainedConfigs,
}: {
  addedConfigs: Config[]
  deletedConfigs: Config[]
  remainedConfigs: Config[]
}): string {
  let result = 'Confirm the following changes:'

  if (addedConfigs.length > 0) {
    result += '\n\n'
    result += 'Will be installed:\n'
    result += chalk.green(addedConfigs.map(configToName).join(', '))
  }

  if (deletedConfigs.length > 0) {
    result += '\n\n'
    result += 'Will be deleted:\n'
    result += chalk.red(deletedConfigs.map(configToName).join(', '))
  }

  if (remainedConfigs.length > 0) {
    result += '\n\n'
    result += 'Will stay unchanged:\n'
    result += chalk.yellow(remainedConfigs.map(configToName).join(', '))
  }

  result += '\n\n'

  return result
}

interface AskQuestionsParams {
  prettierConfigMeta: PrettierConfigMeta
  installedConfigs: Config[]
}

export async function askQuestions({
  prettierConfigMeta,
  installedConfigs,
}: AskQuestionsParams): Promise<Answers> {
  const installedConfigsMap = toMap(installedConfigs)
  const hasInstalledConfigs = installedConfigsMap.size > 0

  const configChoices: ConfigChoice[] = DEFAULT_CONFIG_CHOICES.map(choice => {
    const { name, value } = choice

    const overridedProps: Partial<ConfigChoice> = {}

    if (installedConfigsMap.has(value) && value !== 'base') {
      overridedProps.name = `${name} (installed)`
      overridedProps.checked = true
    }

    return {
      ...choice,
      ...overridedProps,
    }
  })

  const { configs: updatedConfigs } = await inquirer.prompt<{
    configs: Config[]
  }>([
    {
      type: 'checkbox',
      name: 'configs',
      message: createConfigsMessage({ hasInstalledConfigs }),
      choices: configChoices,
      validate: input => {
        if (!input.includes('base')) {
          return 'base config is required'
        }

        return true
      },
    },
  ])

  const addedConfigs = updatedConfigs.filter(config => {
    return !installedConfigs.includes(config)
  })

  const deletedConfigs = installedConfigs.filter(config => {
    return !updatedConfigs.includes(config)
  })

  const remainedConfigs = updatedConfigs.filter(config => {
    return installedConfigs.includes(config)
  })

  const hasChanges = addedConfigs.length > 0 || deletedConfigs.length > 0

  if (hasChanges) {
    const { isConfirmed } = await inquirer.prompt<{ isConfirmed: boolean }>([
      {
        type: 'confirm',
        name: 'isConfirmed',
        message: createConfirmationMessage({
          addedConfigs,
          deletedConfigs,
          remainedConfigs,
        }),
      },
    ])

    if (!isConfirmed) {
      throw new Error('Confirmation had not been received')
    }
  }

  const { shouldAddRecommendedPrettierConfig } = await inquirer.prompt<{
    shouldAddRecommendedPrettierConfig: boolean
  }>([
    {
      type: 'confirm',
      name: 'shouldAddRecommendedPrettierConfig',
      message: 'Add recommended prettier config? (.prettierrc)',
      when: () => {
        if (prettierConfigMeta.isExist) {
          return false
        }

        return addedConfigs.includes('prettier')
      },
      default: true,
    },
  ])

  return {
    updatedConfigs,
    addedConfigs,
    deletedConfigs,
    shouldAddRecommendedPrettierConfig,
  }
}
