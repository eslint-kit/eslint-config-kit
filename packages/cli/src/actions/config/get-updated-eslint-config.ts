import program from 'commander'
import { isFromConfigKit, toConfigName } from '../../lib/config-name-helpers'
import {
  CONFIG_PRIORITIES,
  CONFIG_PREFIX,
  DEPENDENCIES,
} from '../../lib/constants'
import { EslintConfigMeta, Json, Config } from '../../lib/shared-types'

function byPriority(a: string, b: string): number {
  if (!isFromConfigKit(a) || !isFromConfigKit(b)) {
    return 0
  }

  const priorityA = CONFIG_PRIORITIES[toConfigName(a)]
  const priorityB = CONFIG_PRIORITIES[toConfigName(b)]

  return priorityA - priorityB
}

interface GetUpdateConfigParams {
  eslintConfigMeta: EslintConfigMeta
  finalConfigs: Config[]
  useTs: boolean
}

export function getUpdatedEslintConfig({
  eslintConfigMeta,
  finalConfigs,
  useTs,
}: GetUpdateConfigParams): Json {
  const { content } = eslintConfigMeta

  const updatedConfig = Object.assign({}, content)

  if (!updatedConfig.extends) {
    updatedConfig.extends = []
  }

  if (!Array.isArray(updatedConfig.extends)) {
    updatedConfig.extends = [updatedConfig.extends]
  }

  const addedExtends = finalConfigs.map(
    (configName) => CONFIG_PREFIX + configName
  )

  updatedConfig.extends = updatedConfig.extends
    .filter((config) => !isFromConfigKit(config))
    .concat(addedExtends)
    .sort(byPriority)

  updatedConfig.parser = useTs
    ? DEPENDENCIES.TS_PARSER
    : DEPENDENCIES.BABEL_PARSER

  const { workspace } = program.opts()

  if (workspace) {
    updatedConfig.root = true
  }

  return updatedConfig
}
