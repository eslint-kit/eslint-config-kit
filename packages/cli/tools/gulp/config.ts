// All paths are related to the base dir
export const sources = ['lib', 'actions', 'commands', 'bin'];
