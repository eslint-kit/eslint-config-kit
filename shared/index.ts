export * from './constants'
export * from './import'
