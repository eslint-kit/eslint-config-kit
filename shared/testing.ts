import { ESLint, Linter } from "eslint"
import path from 'path'
import fs from 'fs'

export interface Options {
  config: any
  files: string[]
}

export async function testConfig({ config, files }: Options) {
  const cli = new ESLint({
    baseConfig: config as Linter.Config<Linter.RulesRecord>,
    useEslintrc: false,
    cwd: path.resolve(process.cwd(), '../..'),
  })

  for (const file of files) {
    const filePath = path.resolve(process.cwd(), `./tests/${file}`)
    const code = fs.readFileSync(filePath).toString()
    const result = await cli.lintText(code, { filePath })
    expect(result).toMatchSnapshot()
  }
}