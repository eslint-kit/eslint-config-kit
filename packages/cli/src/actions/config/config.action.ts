import chalk from 'chalk'
import { updateEslintConfig } from '../../lib/update-eslint-config'
import { log } from '../../lib/util/log'
import { MESSAGES } from '../../lib/ui/messages'
import { addRecommendedPrettierConfig } from '../../lib/add-recommended-prettier-config'
import { getDependenciesToInstall } from '../../lib/get-dependencies-to-install'
import { updateDependencies } from '../../lib/dependencies-work'
import { getDataBySchema } from '../../lib/get-data-by-schema'
import { getDependenciesToDelete } from '../../lib/get-dependencies-to-delete'
import { getWrongDependencies } from '../../lib/get-wrong-dependencies'
import { getRequiredDependencies } from '../../lib/get-required-dependencies'
import { getWrongDependenciesToUpdate } from '../../lib/get-wrong-dependencies-to-update'
import { getFinalConfigs } from '../../lib/get-final-configs'
import { askQuestions } from './ask-questions'
import { getUpdatedEslintConfig } from './get-updated-eslint-config'
import { LOCAL_MESSAGES } from './ui/local-messages'

export class ConfigAction {
  static async process(): Promise<void> {
    const {
      rootDir,
      packageManager,
      packageJson,
      eslintConfigMeta,
      prettierConfigMeta,
      installedConfigs,
      installedDependencies,
    } = await getDataBySchema({
      rootDir: true,
      packageManager: true,
      packageJson: true,
      eslintConfigMeta: true,
      prettierConfigMeta: true,
      installedConfigs: true,
      installedDependencies: true,
    })

    const {
      updatedConfigs,
      shouldAddRecommendedPrettierConfig,
    } = await askQuestions({
      prettierConfigMeta,
      installedConfigs,
    })

    const { useTs } = await getDataBySchema(
      {
        useTs: true,
      },
      { installedDependencies, installedConfigs },
      { updatedConfigs }
    )

    const finalConfigs = getFinalConfigs({
      packageJson,
      updatedConfigs,
    })

    const updatedConfig = getUpdatedEslintConfig({
      eslintConfigMeta,
      finalConfigs,
      useTs,
    })

    await updateEslintConfig({
      rootDir,
      updatedConfig,
      packageJson,
      eslintConfigMeta,
    })

    if (shouldAddRecommendedPrettierConfig) {
      addRecommendedPrettierConfig({ rootDir })
    }

    const requiredDependencies = getRequiredDependencies({
      installedDependencies,
      finalConfigs,
      useTs,
    })

    const wrongDependencies = getWrongDependencies({
      packageJson,
      requiredDependencies,
    })

    const wrongDependenciesToUpdate = await getWrongDependenciesToUpdate({
      wrongDependencies,
    })

    const dependenciesToInstall = getDependenciesToInstall({
      requiredDependencies,
      installedDependencies,
      wrongDependenciesToUpdate,
    })

    const dependenciesToDelete = getDependenciesToDelete({
      requiredDependencies,
      installedDependencies,
      wrongDependenciesToUpdate,
    })

    await updateDependencies({
      packageManager,
      dependenciesToInstall,
      dependenciesToDelete,
    })

    log(LOCAL_MESSAGES.FINISHED, chalk.green, [false, true])
    log(MESSAGES.PLEASE_RESTART, [chalk.magenta, chalk.bold], [false, true])
  }

  public static async handle(): Promise<void> {
    try {
      await ConfigAction.process()
    } catch (err) {
      log(err.message, chalk.red)
    }
  }
}
