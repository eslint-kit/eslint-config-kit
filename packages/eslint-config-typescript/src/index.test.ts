import { testConfig } from '../../../shared/testing'
import config from './index'

describe('@eslint-kit/typescript', () => {
  it('should match snapshots', async () => {
    await testConfig({
      config: {
        ...config,
        parser: '@typescript-eslint/parser',
      },
      files: ['basic-clear', 'basic-warn', 'basic-error'],
      extension: 'ts',
    })
  })
})
