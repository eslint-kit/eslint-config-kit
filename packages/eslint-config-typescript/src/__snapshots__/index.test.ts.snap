// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@eslint-kit/typescript should match snapshots 1`] = `
Array [
  Object {
    "errorCount": 0,
    "filePath": "/Users/risenforces/Desktop/Projects/eslint-kit/packages/eslint-config-typescript/tests/basic-clear.ts",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`@eslint-kit/typescript should match snapshots 2`] = `
Array [
  Object {
    "errorCount": 0,
    "filePath": "/Users/risenforces/Desktop/Projects/eslint-kit/packages/eslint-config-typescript/tests/basic-warn.ts",
    "fixableErrorCount": 0,
    "fixableWarningCount": 1,
    "messages": Array [
      Object {
        "column": 6,
        "endColumn": 10,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            0,
            12,
          ],
          "text": "interface Test ",
        },
        "line": 1,
        "message": "Use an \`interface\` instead of a \`type\`.",
        "messageId": "interfaceOverType",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/consistent-type-definitions",
        "severity": 1,
      },
      Object {
        "column": 6,
        "endColumn": 10,
        "endLine": 1,
        "line": 1,
        "message": "'Test' is defined but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/no-unused-vars",
        "severity": 1,
      },
    ],
    "source": "type Test = {
  one: number
}
",
    "usedDeprecatedRules": Array [],
    "warningCount": 2,
  },
]
`;

exports[`@eslint-kit/typescript should match snapshots 3`] = `
Array [
  Object {
    "errorCount": 1,
    "filePath": "/Users/risenforces/Desktop/Projects/eslint-kit/packages/eslint-config-typescript/tests/basic-error.ts",
    "fixableErrorCount": 1,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 12,
        "endColumn": 18,
        "endLine": 1,
        "fix": Object {
          "range": Array [
            11,
            17,
          ],
          "text": "number",
        },
        "line": 1,
        "message": "Don't use \`Number\` as a type. Use number instead",
        "messageId": "bannedTypeMessage",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/ban-types",
        "severity": 2,
      },
    ],
    "source": "const one: Number = 1
export { one }
",
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;
